modules/shBidAdapter.js

```diff
@@ -154,25 +176,95 @@ function createBids(bidRes, reqData) {
     }
     if (reqBid.mediaType === BANNER) {
       bidUnit.ad = getBannerHtml(bid, reqBid, reqData);
+    } else if (reqBid.context === 'outstream') {
+      const renderer = Renderer.install({
+        id: bid.bidId,
+        url: '//',
+        config: {
+          playerId: reqBid.playerId,
+          width: bid.video.width,
+          height: bid.video.height,
+          vastUrl: bid.vastTag,
+          vastXml: bid.vastXml,
+          debug: reqData.debug,
+          isStage: !!reqData.meta.stage,
+          customRender: utils.getBidIdParameter('customRender', currentBidRequest.params.outstreamOptions),
+          slot: utils.getBidIdParameter('slot', currentBidRequest.params.outstreamOptions),
+          iframe: utils.getBidIdParameter('iframe', currentBidRequest.params.outstreamOptions),
+        }
+      });
+      renderer.setRender(outstreamRender);
+      bidUnit.renderer = renderer;
     }
     bids.push(bidUnit);
   });
 
   return bids;
 }
 
+function outstreamRender(bid) {
+  const embedCode = createOutstreamEmbedCode(bid);
+  if (typeof bid.renderer.config.customRender === 'function') {
+    bid.renderer.config.customRender(bid, embedCode);
+  } else {
+    try {
+      const inIframe = utils.getBidIdParameter('iframe', bid.renderer.config);
+      if (inIframe && window.document.getElementById(inIframe).nodeName === 'IFRAME') {
```

Comment:
How should this work with DFP?

The **iframe** option is hard, because the iframe ID is assigned by `gpt.js`, which can change the naming scheme at any time, breaking the integration.

The **slot** option doesn't work either as DFP creatives are always delivered in an iframe and the standard prebid creative doesn't have any div elements.

From other implementations there are better options:

## inFrame flag

If you really need to know if you are in an iFrame or not use a flag instead. Then append the necessary code to the `body` and `head`. Scripts, divs, whatever you want. Not need to break out and search for the iFrame.

## slot_id

If you want to break out of the iFrame, then try to do something like `window.top.document.getElementById(..)`. I guess prebid has some helper functions that prevent illegal access if the creative was not delivered in a friendly iframe.

Reply:
These options are not for DFP integration. They need for our publishers, which want to set our outstream directly on their page. They can use id of existing iframe or slot id (can be used any html element).

Reply:
tl;dr IMHO the `slot_id` and `iframe` parameter are unnecessary for an integration that relies on an Ad Server such as DFP that delivers creatives in iframes. From my limited experience it's a safe guess that the player was delivered in an iframe through and ad server. Especially as a Prebid module.

> These options are not for DFP integration. They need for our publishers, which want to set our outstream directly on their page. They can use id of existing iframe or slot id (can be used any html element).

I'm not sure how prevalent DFP is as an Ad Server, but I guess there's a fair amount of publisher that use DFP as an Ad Server along with Prebid. 

Your solution is what is documented as [Option 2: Serving without an ad server](http://prebid.org/dev-docs/show-outstream-video-ads.html#option-2-serving-without-an-ad-server) in the outstream-video-ads documentation. We haven't heard of any publisher doing this.

> If you would like to integrate our outstream with DFP, please, give me a link with an example of your bidding. We will try to found a solution.

This is not necessarily DFP relevant. Delivering ad creatives in an iFrame is IMHO a good practice to avoid ad fraud and illegal page access by creatives.

My point here is that the implementation is unnecessary complex and hinders the integration with any Ad Server that delivers in iFrames. At least this is how I understand the code :wink: . Please correct me if I missed something obvious.

Here's the [example page from prebid outstream working examples section](http://prebid.org/dev-docs/show-outstream-video-ads.html#working-examples):

http://prebid.org/dev-docs/show-outstream-video-ads.html

The appNexus outstream renders everything inside the iframe and only adds a small `div` for handling the expansion of the video. As mentioned in the initial issue this is how the appNexus adapter does it:
 [AppNexus Adpater `newRenderer`](https://github.com/prebid/Prebid.js/blob/master/modules/appnexusBidAdapter.js#L428-L452) and [AppNexus Adapter `oustreamRenderer`]

Reply:
Hi @muuki88! We created a demo page how we can work with DFP http://test.showheroes.com/schema/sh-prebid-outstream-google 
Please, take a look. We will return "banner" type and you can use "renderAd" function inside DFP. In this case, no need to use the frame wrapper for custom frames and slots.

Reply:
Hi @veranevera :smiley: 

Thanks a lot for the example page :+1: The banner solutions sounds good to me. Still unsure about the use case of the other options in Prebid, but you know what you are doing :grin: 

I'll give it a try and keep you posted. I saw that you were using `1x1` as a size, so the player resizes itself, right?

Reply:
@muuki88 we are looking forward 💃 
The player from the demo is responsive :) do you need fixed sizes?

Reply:
Nope. Responsive is perfect :smile: